lib/c_glib/src/thrift/c_glib/transport/thrift_ssl_socket.c

```diff
@@ -0,0 +1,772 @@
+	  fprintf(stderr,"The error returned was %d\n", SSL_get_error(ssl_socket->ssl, rc));
+	  thrift_ssl_socket_get_error(error, "Not possible to connect", THRIFT_SSL_SOCKET_ERROR_CIPHER_NOT_AVAILABLE);
+	  return FALSE;
+      }
+  }else
+    return FALSE;
+
+  return thrift_ssl_socket_authorize(transport, error);
+}
+
+gboolean
+thrift_ssl_socket_create_ssl_context(ThriftTransport * transport, GError **error)
+{
+  ThriftSSLSocket *socket = THRIFT_SSL_SOCKET (transport);
+
+  if(socket->ctx!=NULL){
+      if(socket->ssl!=NULL) {
+	  return TRUE;
+      }
+
+      socket->ssl = SSL_new(socket->ctx);
+      if (socket->ssl == NULL) {
+	  g_set_error (error, THRIFT_TRANSPORT_ERROR,
+		       THRIFT_SSL_SOCKET_ERROR_TRANSPORT,
+		       "Unable to create SSL context");
+	  return FALSE;
+      }
+  }
+
+  return TRUE;
+}
+
+/**
+ *
+ * @param ssl_socket The ssl socket
+ * @param file_name The file name of the PEM certificate chain
+ * @return
+ */
+gboolean thrift_ssl_load_cert_from_file(ThriftSSLSocket *ssl_socket, const char *file_name)
+{
+  char error_buffer[255];
+  if (!thrift_ssl_socket_openssl_initialized) {
+      g_error("OpenSSL is not initialized yet");
+      return FALSE;
+  }
+  int rc = SSL_CTX_load_verify_locations(ssl_socket->ctx, file_name, NULL);
+  if (rc != 1) { /*verify authentication result*/
+      ERR_error_string_n(ERR_get_error(), error_buffer, 254);
+      g_warning("Load of certificates failed: %s!", error_buffer);
+      return FALSE;
+  }
+  return TRUE;
+}
+
+/**
+ * Load a certificate chain from memory
+ * @param ssl_socket the ssl socket
+ * @param chain_certs the buffer to load PEM from
+ * @return
+ */
+gboolean thrift_ssl_load_cert_from_buffer(ThriftSSLSocket *ssl_socket, const char chain_certs[])
+{
+  gboolean retval = FALSE;
+  /* Load chain of certs*/
+  X509 *cacert=NULL;
+  BIO *mem = BIO_new_mem_buf(chain_certs,strlen(chain_certs));
+  X509_STORE *cert_store = SSL_CTX_get_cert_store(ssl_socket->ctx);
+
+  if(cert_store!=NULL){
+      int index = 0;
+      while ((cacert = PEM_read_bio_X509(mem, NULL, 0, NULL))!=NULL) {
+	  if(cacert) {
+	      g_debug("Our certificate name is %s", cacert->name);
+	      X509_STORE_add_cert(cert_store, cacert);
+	      X509_free(cacert);
+	      cacert=NULL;
+	  } /* Free immediately */
+	  index++;
+      }
+      retval=TRUE;
+  }
+  BIO_free(mem);
+  return retval;
+}
+
+gboolean
+thrift_ssl_socket_authorize(ThriftTransport * transport, GError **error)
+{
+  ThriftSocket *socket = THRIFT_SOCKET (transport);
+  ThriftSSLSocket *ssl_socket = THRIFT_SSL_SOCKET (transport);
+  ThriftSSLSocketClass *cls = THRIFT_SSL_SOCKET_GET_CLASS(ssl_socket);
+  gboolean authorization_result = FALSE;
+  /* We still don't support it */
+  if(cls!=NULL && ssl_socket->ssl!=NULL){
+      int rc = SSL_get_verify_result(ssl_socket->ssl);
+      if (rc != X509_V_OK) { /* verify authentication result */
+	  if(rc==X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT && ssl_socket->allow_selfsigned){
```

Comment:
The selfsigned verification can be also skipped by doing other kind of initialization in openssl. This is why we provide our own function. To make sure it will enforce the check. So I'm not happy with it but it's necessary for now. Hope we can remove in the near future.